cmd/influxd/inspect/report_tsm/report_tsm.go

```diff
@@ -0,0 +1,356 @@
+		detailed:           a.detailed,
+		tagCardinalities:   tagCardinalities,
+		measCardinalities:  measCardinalities,
+		fieldCardinalities: fieldCardinalities,
+		dbCardinalities:    dbCardinalities,
+	}.printSummary(cmd)
+
+	cmd.Printf("Completed in %s\n", time.Since(start))
+	return nil
+}
+
+type printArgs struct {
+	fileCount        int
+	minTime, maxTime int64
+	estTitle         string
+	totalSeries      counter
+	detailed         bool
+
+	tagCardinalities   map[string]counter
+	measCardinalities  map[string]counter
+	fieldCardinalities map[string]counter
+	dbCardinalities    map[string]counter
+}
+
+func (p printArgs) printSummary(cmd *cobra.Command) {
+	cmd.Printf("\nSummary:")
+	cmd.Printf("  Files: %d\n", p.fileCount)
+	cmd.Printf("  Time Range: %s - %s\n",
+		time.Unix(0, p.minTime).UTC().Format(time.RFC3339Nano),
+		time.Unix(0, p.maxTime).UTC().Format(time.RFC3339Nano),
+	)
+	cmd.Printf("  Duration: %s \n\n", time.Unix(0, p.maxTime).Sub(time.Unix(0, p.minTime)))
+
+	cmd.Printf("Statistics\n")
+	cmd.Printf("  Series:\n")
+	for db, counts := range p.dbCardinalities {
+		cmd.Printf("     - %s%s: %d (%d%%)\n", db, p.estTitle, counts.Count(), int(float64(counts.Count())/float64(p.totalSeries.Count())*100))
+	}
+	cmd.Printf("  Total%s: %d\n", p.estTitle, p.totalSeries.Count())
+
+	if p.detailed {
+		cmd.Printf("\n  Measurements (est):\n")
+		for _, t := range sortKeys(p.measCardinalities) {
+			cmd.Printf("    - %v: %d (%d%%)\n", t, p.measCardinalities[t].Count(), int((float64(p.measCardinalities[t].Count())/float64(p.totalSeries.Count()))*100))
+		}
+
+		cmd.Printf("\n  Fields (est):\n")
+		for _, t := range sortKeys(p.fieldCardinalities) {
+			cmd.Printf("    - %v: %d\n", t, p.fieldCardinalities[t].Count())
+		}
+
+		cmd.Printf("\n  Tags (est):\n")
+		for _, t := range sortKeys(p.tagCardinalities) {
+			cmd.Printf("    - %v: %d\n", t, p.tagCardinalities[t].Count())
+		}
+	}
+}
+
+// sortKeys is a quick helper to return the sorted set of a map's keys
+func sortKeys(vals map[string]counter) (keys []string) {
+	for k := range vals {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	return keys
+}
+
+func (a *args) walkShardDirs(root string, fn func(db, rp, id, path string) error) error {
+	type location struct {
+		db, rp, id, path string
+	}
+
+	var dirs []location
+	if err := filepath.Walk(root, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if info.IsDir() {
+			return nil
+		}
+
+		if filepath.Ext(info.Name()) == "."+tsm1.TSMFileExtension {
+			shardDir := filepath.Dir(path)
+
+			if err := a.isShardDir(shardDir); err != nil {
+				return err
+			}
+			absPath, err := filepath.Abs(path)
+			if err != nil {
+				return err
+			}
+			parts := strings.Split(absPath, string(filepath.Separator))
+			db, rp, id := parts[len(parts)-4], parts[len(parts)-3], parts[len(parts)-2]
+			dirs = append(dirs, location{db: db, rp: rp, id: id, path: path})
+			return nil
```

Comment:
Will this algorithm result in duplicate entries when a shard contains multiple TSM files? Could we guard against that?

Reply:
Looking a 2nd time, I see `path` in the location struct, so there shouldn't be duplicates